aws-java-sdk-sts/src/main/java/com/amazonaws/auth/STSSessionCredentialsProvider.java

```diff
@@ -119,36 +160,119 @@ public STSSessionCredentialsProvider(AWSCredentialsProvider longLivedCredentials
      */
     public void setSTSClientEndpoint(String endpoint) {
         securityTokenService.setEndpoint(endpoint);
-        sessionCredentials = null;
+        sessionCredentials.set(null);
     }
 
     @Override
     public AWSSessionCredentials getCredentials() {
         if (needsNewSession()) startSession();
 
-        return sessionCredentials;
+        return sessionCredentials.get().sessionCredentials;
     }
 
     @Override
     public void refresh() {
-        startSession();
+        asyncNewSession();
```

Comment:
asyncNewSession() is thread-safe and lock-free, multiple threads can concurrently call refresh() without side effects. All synchronized would do here is prevent liveness.